### PR TITLE
[don't auto-bump] Crypto algo alignment

### PR DESCRIPTION
**Pull request by toly-kournik**
* **Problem Overview**  
  < a few sentences about what problem we want to solve with our change(e.g. description of the bug we are fixing) >
* **Testing Done**  
  < describe automated and/or manual test scenarios >

### CODE REVIEW

File: threshsign/src/eddsa/EdDSAMultisigFactory.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include "threshsign/eddsa/EdDSAMultisigFactory.h"
+#include "crypto/openssl/crypto.hpp"
+#include "threshsign/eddsa/EdDSAThreshsignKeys.h"
+#include "threshsign/eddsa/EdDSAMultisigSigner.h"
+#include "threshsign/eddsa/EdDSAMultisigVerifier.h"
+#include <boost/algorithm/hex.hpp>
+#include <iostream>
+#include "crypto/openssl/crypto.hpp"
```
Comment (glevkovich) [inline threshsign/src/eddsa/EdDSAMultisigFactory.cpp:21]: can you please seperate std library includes, 3rd party includes and own code includes (with empty line between)?

File: threshsign/src/eddsa/EdDSAMultisigSigner.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  LOG_DEBUG(EDDSA_MULTISIG_LOG, KVLOG(id_));
+  SingleEdDSASignature result;
+  auto outSigBytesLen = result.signatureBytes.size();
+  signBuffer(reinterpret_cast<const concord::Byte *>(hash), static_cast<size_t>(hashLen), result.signatureBytes.data());
+  ConcordAssertEQ(outSigBytesLen, result.signatureBytes.size());
+  result.id = id_;
+  std::memcpy(outSig, &result, sizeof(SingleEdDSASignature));
+}
+const IShareSecretKey &EdDSAMultisigSigner::getShareSecretKey() const { return privateKey_; }
+const IShareVerificationKey &EdDSAMultisigSigner::getShareVerificationKey() const { return publicKey_; }
```
Comment (glevkovich) [inline threshsign/src/eddsa/EdDSAMultisigSigner.cpp:34]: line break at the end.

File: threshsign/src/eddsa/EdDSAMultisigVerifier.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  const SingleEdDSASignature *allSignatures = reinterpret_cast<const SingleEdDSASignature *>(sig);
+  size_t validSignatureCount = 0;
+
+  for (int i = 0; i < static_cast<int>(signatureCountInBuffer); i++) {
+    auto &currentSignature = allSignatures[i];
+    if (currentSignature.id == 0 || currentSignature.id >= verifiers_.size()) {
+      LOG_ERROR(EDDSA_MULTISIG_LOG, "Invalid signer id" << KVLOG(currentSignature.id, verifiers_.size()));
+      continue;
```
Comment (glevkovich) [inline threshsign/src/eddsa/EdDSAMultisigVerifier.cpp:110]: Is this a valid case to just continue? I am not "in the loop" on where is the signer id coming from, if it comes from network/storage then we shouldnt terminate - else everyone who is byzantine autehticated can crash the replica. if it's some internal error maybe termination is better.

File: util/include/hex_tools.h [hunk 4]
```diff
@@ -37,4 +37,9 @@ std::string sliverToHex(const Sliver &sliver);
 // Converts a byte vector into a hex string
 std::string vectorToHex(const std::vector<std::uint8_t> &data);
 
+// Converts a hex string into a char vector. Handles leading 0x characters (if present).
+// Input hex: "616263" or "0x616263"
+// Returns: abc
+std::vector<uint8_t> unhex(const std::string &hex);
```
Comment (glevkovich) [inline util/include/hex_tools.h:43]: hextoVector?
this name is not a good choice i think..

File: util/include/macros.h [hunk 5]
```diff
@@ -179,3 +179,5 @@
 #define KVARGS3(_1, _2, _3) (#_1, _1, #_2, _2, #_3, _3)
 #define KVARGS2(_1, _2) (#_1, _1, #_2, _2)
 #define KVARGS1(_1) (#_1, _1)
+
+#define UNUSED(expr) (void)(expr)
```
Comment (glevkovich) [inline util/include/macros.h:183]: https://en.cppreference.com/w/cpp/language/attributes/maybe_unused

File: util/include/types.hpp [hunk 6]
```diff
@@ -1,20 +1,18 @@
...
 
-#include <ostream>
-
-std::ostream& operator<<(std::ostream& out, const IPublicKey& pk) {
-  out << pk.toString();
-  return out;
-}
+namespace concord {
+using Byte = uint8_t;
+}
```
Comment (glevkovich) [inline util/include/types.hpp:18]: new line

File: util/src/crypto/crypto.cpp [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+#include "crypto/crypto.hpp"
+#include "Logger.hpp"
+#include "assertUtils.hpp"
+#include "hex_tools.h"
+#include "string.hpp"
+#include "types.hpp"
+#include "crypto/openssl/EdDSA.hpp"
+#include "util/filesystem.hpp"
+#include <utility>
```
Comment (glevkovich) [inline util/src/crypto/crypto.cpp:22]: can seperate std includes?

File: util/src/openssl/certificates.cpp [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+                       const string& cert_root_directory,
+                       uint32_t& remote_peer_id,
+                       string& conn_type,
+                       bool use_unified_certs) {
+  // First get the source ID
+  static constexpr size_t SIZE = 512;
+  string subject(SIZE, 0);
+  X509_NAME_oneline(X509_get_subject_name(&cert_to_verify), subject.data(), SIZE);
+
+  int peerIdPrefixLength = 3;
```
Comment (glevkovich) [inline util/src/openssl/certificates.cpp:134]: A general comment about coding style:
Why some of the code is snake-case and some camel-case?
I know that if I modify an existing code i tend to keep the existing style.
But this is a new file and it mixes between 2 styles...looks strange to me.
e.g peerIdPrefixLengthand vs origin_cert_path

**Review by eyalrund — APPROVED** [review]

**Comment by cjn** [comment]
@glevkovich - The review comments if any unaddressed will be taken care as part of separate PR as we may need to merge this PR to proceed. Kindly approve.
cc: @eyalrund

**Review by glevkovich — APPROVED** [review]
I'm approving, but none of my comments have been addressed & the number of reviewers is too low for such a large change.

File: bftengine/src/bftengine/InternalBFTClient.hpp [hunk 9]
```diff
@@ -28,6 +28,7 @@ class IInternalBFTClient {
   // Note: users should be aware that if they push a request from the consumer/replica thread(s), the given callback
   // will be called in the same thread.
   virtual uint64_t sendRequest(uint64_t flags, uint32_t requestLength, const char* request, const std::string& cid) = 0;
+  uint64_t sendRequest(uint64_t flags, const concord::serialize::Serializable& msg, const std::string& cid);
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/src/bftengine/InternalBFTClient.hpp:31]: No provision to convert into `const` functions.

File: bftengine/src/bftengine/SigManager.cpp [hunk 10]
```diff
@@ -237,22 +244,24 @@ bool SigManager::verifySig(
   return result;
 }
 
-void SigManager::sign(const char* data, size_t dataLength, char* outSig, uint16_t outSigLength) const {
-  std::string str_data(data, dataLength);
-  std::string sig;
-  sig = mySigner_->sign(str_data);
-  outSigLength = sig.size();
-  std::memcpy(outSig, sig.c_str(), outSigLength);
+size_t SigManager::sign(const concord::Byte* data, size_t dataLength, concord::Byte* outSig) const {
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/src/bftengine/SigManager.cpp:247]: 2. `outSigLength` was redundant.

File: util/include/hex_tools.h (hunk 4 above)
Comment (pkthapa) [reply util/include/hex_tools.h:43]: Trying to match with `boost::algorithm::unhex()`.

File: bftengine/src/bcstatetransfer/RangeValidationTree.hpp [hunk 11]
```diff
@@ -150,29 +152,52 @@ class RangeValidationTree {
...
+      val_ = (val_ - v.val_) % getModulo();
+      return *this;
+    }
+    bool operator!=(const NodeVal& v) const { return (val_ != v.val_); }
+    bool operator==(const NodeVal& v) const { return (val_ == v.val_); }
+
+    void setNegative() { val_.setNegative(); }
+
+    std::string toString() const noexcept { return val_.toHexString(); }
+    std::string toHexString() const noexcept { return val_.toHexString(); }
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/src/bcstatetransfer/RangeValidationTree.hpp:184]: Resolved https://github.com/vmware/concord-bft/pull/2851

File: bftengine/include/bftengine/KeyExchangeMsg.hpp [hunk 12]
```diff
@@ -22,11 +22,13 @@ struct KeyExchangeMsg : public concord::serialize::SerializableFactory<KeyExchan
   uint16_t repID;
   uint64_t generated_sn;
   uint64_t epoch;
+  uint16_t algorithm;
 
   std::string toString() const {
     std::stringstream ss;
     ss << "pubkey [" << pubkey << "] replica id [" << repID << "] generate sequence number [" << generated_sn
-       << "] epoch [" << epoch << "]";
+       << "] epoch [" << epoch << "]"
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/include/bftengine/KeyExchangeMsg.hpp:30]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: bftengine/src/bftengine/CollectorOfThresholdSignatures.hpp [hunk 13]
```diff
@@ -50,6 +51,10 @@ class CollectorOfThresholdSignatures {
     RepInfo info = {partialSigMsg, SigState::Unknown};
     replicasInfo[repId] = info;
 
+    /*if ((((InternalReplicaApi*)this->context)->getReplicaConfig().numReplicas) == 1) {
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/src/bftengine/CollectorOfThresholdSignatures.hpp:54]: Removed. https://github.com/vmware/concord-bft/pull/2851

File: bftengine/src/bftengine/KeyExchangeManager.cpp [hunk 14]
```diff
@@ -224,51 +236,37 @@ void KeyExchangeManager::sendMainPublicKey() {
...
-  auto [prv, pub] = multiSigKeyHdlr_->generateMultisigKeyPair();
-  candidate_private_keys_.generated.priv = prv;
-  candidate_private_keys_.generated.pub = pub;
-  candidate_private_keys_.generated.cid = cid;
-  candidate_private_keys_.generated.sn = sn;
-
-  LOG_INFO(KEY_EX_LOG, "Sending key exchange :" << KVLOG(cid, pub));
-  msg.pubkey = pub;
+  auto& candidate = candidate_private_keys_.generated;
+  bool generateNewPair = sn != candidate.sn || candidate.cid.empty();
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/src/bftengine/KeyExchangeManager.cpp:248]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: bftengine/src/bftengine/NullStateTransfer.cpp [hunk 15]
```diff
@@ -63,7 +63,8 @@ void NullStateTransfer::getDigestOfCheckpoint(uint64_t checkpointNumber,
   LOG_WARN(GL, "State digest is only based on sequence number (because state transfer module has not been loaded)");
 
   Digest d;
-  DigestUtil::compute((char*)&checkpointNumber, sizeof(checkpointNumber), (char*)&d, sizeof(d));
+  DigestGenerator digestGenerator;
+  digestGenerator.compute((char*)&checkpointNumber, sizeof(checkpointNumber), (char*)&d, sizeof(d));
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/src/bftengine/NullStateTransfer.cpp:67]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: bftengine/src/bftengine/ReadOnlyReplica.cpp [hunk 16]
```diff
@@ -12,6 +12,7 @@
 #include <bftengine/Replica.hpp>
 #include <optional>
 #include <functional>
+#include <bitset>
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/src/bftengine/ReadOnlyReplica.cpp:15]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 17]
```diff
@@ -4164,15 +4169,16 @@ ReplicaImp::ReplicaImp(const LoadedReplicaData &ld,
           throw;
         }
 
-        Digest d;
-        Digest::digestOfDigest(e.getPrePrepareMsg()->digestOfRequests(), d);
+        Digest digest;
+        e.getPrePrepareMsg()->digestOfRequests().digestOfDigest(digest);
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply bftengine/src/bftengine/ReplicaImp.cpp:4212]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: kvbc/src/pruning_handler.cpp [hunk 18]
```diff
@@ -16,31 +16,35 @@
...
+  auto actualSignatureLength = signer_->sign(ser, signature.data());
+  ConcordAssertEQ(actualSignatureLength, signer_->signatureLength());
+  block.signature = std::move(signature);
 }
 
-RSAPruningSigner::RSAPruningSigner(const std::string& key)
-    : signer_{std::make_unique<concord::util::crypto::RSASigner>(
-          key, concord::util::crypto::KeyFormat::HexaDecimalStrippedFormat)} {}
+PruningSigner::PruningSigner(const std::string& key) {
+  signer_ = concord::crypto::Factory::getSigner(key, ReplicaConfig::instance().replicaMsgSigningAlgo);
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply kvbc/src/pruning_handler.cpp:37]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: kvbc/tools/object_store_utility/integrity_checker.cpp [hunk 19]
```diff
@@ -75,7 +76,7 @@ std::pair<BlockId, Digest> IntegrityChecker::getLatestsCheckpointDescriptor() co
   auto it = dynamic_cast<s3::Client*>(s3_dbset_.metadataDBClient.get())
                 ->getIterator<s3::Client::SortByModifiedDescIterator>();
   auto [key, val] = it->seekAtLeast(Sliver::copy(checkpoints_prefix_.data(), checkpoints_prefix_.length()));
-  (void)val;
+  UNUSED(val);
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply kvbc/tools/object_store_utility/integrity_checker.cpp:79]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: secretsmanager/src/aes.cpp [hunk 20]
```diff
@@ -14,31 +14,78 @@
...
-  std::vector<uint8_t> cipher;
-  CryptoPP::StringSource ss(
-      input, true, new CryptoPP::StreamTransformationFilter(enc, new CryptoPP::VectorSink(cipher)));
-  return cipher;
-}
-std::string AES_CBC::decrypt(const std::vector<uint8_t>& cipher) {
-  std::string pt;
-  CryptoPP::VectorSource ss(cipher, true, new CryptoPP::StreamTransformationFilter(dec, new CryptoPP::StringSink(pt)));
-  return pt;
+    unique_ptr<unsigned char[]> ciphertext(new unsigned char[input.size() + AES_BLOCK_SIZE]);
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply secretsmanager/src/aes.cpp:34]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: secretsmanager/src/aes.cpp [hunk 21]
```diff
@@ -14,31 +14,78 @@
...
+    unique_ptr<unsigned char[]> plaintext(new unsigned char[input.size() + 1]);
+
+    for (size_t i{0UL}; i < input.size(); ++i) {
+      plaintext.get()[i] = (unsigned char)input[i];
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply secretsmanager/src/aes.cpp:38]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: secretsmanager/src/aes.cpp [hunk 22]
```diff
@@ -14,31 +14,78 @@
...
+
+    ConcordAssert(OPENSSL_SUCCESS ==
+                  EVP_EncryptInit_ex(ctx.get(), EVP_aes_256_cbc(), nullptr, params_.key.data(), params_.iv.data()));
+    ConcordAssert(OPENSSL_SUCCESS ==
+                  EVP_EncryptUpdate(ctx.get(), ciphertext.get(), &c_len, plaintext.get(), input.size()));
+    ConcordAssert(OPENSSL_SUCCESS == EVP_EncryptFinal_ex(ctx.get(), ciphertext.get() + c_len, &f_len));
+
+    const int encryptedMsgLen = c_len + f_len;
+    vector<uint8_t> cipher(encryptedMsgLen);
+    memcpy(&cipher[0], ciphertext.get(), encryptedMsgLen);
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply secretsmanager/src/aes.cpp:55]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: secretsmanager/src/base64.cpp [hunk 23]
```diff
@@ -14,27 +14,75 @@
...
 
+  vector<uint8_t> decodedOutput(calcDecodeLength(b64msg.data()));
+
+  BIO* bio = BIO_new_mem_buf(b64msg.data(), -1);
+  BIO* b64 = BIO_new(BIO_f_base64());
+  bio = BIO_push(b64, bio);
+
+  const int outputLen = BIO_read(bio, decodedOutput.data(), b64msg.size());
+  vector<uint8_t> dec(outputLen);
+  memcpy(&dec[0], decodedOutput.data(), outputLen);
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply secretsmanager/src/base64.cpp:57]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: threshsign/include/threshsign/eddsa/EdDSAMultisigVerifier.h [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+// LICENSE file.
+#pragma once
+#include "threshsign/IThresholdVerifier.h"
+#include "SingleEdDSASignature.h"
+#include "crypto/openssl/EdDSAVerifier.hpp"
+#include "EdDSAThreshsignKeys.h"
+
+class EdDSAMultisigVerifier;
+
+bool isSignatureValid(const SingleEdDSASignature &signature);
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply threshsign/include/threshsign/eddsa/EdDSAMultisigVerifier.h:20]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: threshsign/include/threshsign/eddsa/EdDSAMultisigVerifier.h [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+  const IShareVerificationKey &getShareVerificationKey(ShareID signer) const override;
+
+  bool verifySingleSignature(const concord::Byte *msg, size_t msgLen, const SingleEdDSASignature &signature) const;
+  ~EdDSAMultisigVerifier() override = default;
+
+ private:
+  std::vector<SingleVerifier> verifiers_;
+  const size_t signersCount_;
+  const size_t threshold_;
+};
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply threshsign/include/threshsign/eddsa/EdDSAMultisigVerifier.h:69]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: threshsign/include/threshsign/eddsa/EdDSAThreshsignKeys.h [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+      : concord::crypto::openssl::EdDSAPrivateKey(arr) {}
+  std::string toString() const override { return toHexString(); }
+};
+
+class EdDSAThreshsignPublicKey : public IShareVerificationKey, public concord::crypto::openssl::EdDSAPublicKey {
+ public:
+  EdDSAThreshsignPublicKey(const EdDSAThreshsignPublicKey::ByteArray& arr)
+      : concord::crypto::openssl::EdDSAPublicKey(arr) {}
+  std::string toString() const override { return toHexString(); }
+};
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply threshsign/include/threshsign/eddsa/EdDSAThreshsignKeys.h:29]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: threshsign/include/threshsign/eddsa/SingleEdDSASignature.h [hunk 27]
```diff
@@ -1,25 +1,24 @@
...
-
-}  // namespace util
-}  // namespace concord
+/**
+ * Used as an interface between signers and verifiers
+ */
+struct SingleEdDSASignature {
+  uint64_t id;
+  std::array<concord::Byte, concord::crypto::Ed25519SignatureByteSize> signatureBytes;
+};
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply threshsign/include/threshsign/eddsa/SingleEdDSASignature.h:24]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: threshsign/src/eddsa/EdDSAMultisigFactory.cpp (hunk 1 above)
Comment (pkthapa) [reply threshsign/src/eddsa/EdDSAMultisigFactory.cpp:21]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: threshsign/src/eddsa/EdDSAMultisigSigner.cpp (hunk 2 above)
Comment (pkthapa) [reply threshsign/src/eddsa/EdDSAMultisigSigner.cpp:34]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: util/include/macros.h (hunk 5 above)
Comment (pkthapa) [reply util/include/macros.h:183]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: util/src/crypto/crypto.cpp (hunk 7 above)
Comment (pkthapa) [reply util/src/crypto/crypto.cpp:22]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: util/src/openssl/certificates.cpp (hunk 8 above)
Comment (pkthapa) [reply util/src/openssl/certificates.cpp:134]: Resolved. https://github.com/vmware/concord-bft/pull/2851

File: install_deps.sh [hunk 28]
```diff
@@ -69,7 +69,7 @@ install_third_party_libraries() {
         libz-dev \
         libzstd-dev
 
-    pip3 install --upgrade wheel && pip3 install --upgrade trio
+    pip3 install --upgrade wheel && pip3 install --upgrade trio && pip3 install --upgrade pip
```
[the remark this answers is not in the document]
Comment (pkthapa) [reply install_deps.sh:72]: Will be addressed in https://github.com/vmware/concord-bft/pull/2851

File: util/include/types.hpp (hunk 6 above)
Comment (pkthapa) [reply util/include/types.hpp:18]: Resolved. https://github.com/vmware/concord-bft/pull/2851

**Pull request merged**